ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -117,11 +117,56 @@ As a third-party application developer that uses shipwright under the hood, I wa
 
 ### Implementation Notes
 
-`tbd in the next phase of the proposal`
+
+This ship proposes two changes in the API that are implemented in `buildrun_types.go` and thus propagated in the CRD. First, the BuildRun's status gets the new field `errorDetails` that embeds the `failedAt` field:
+
+```go
+type FailureDetails struct {
+    Reason   string    `json:"reason,omitempty"`
+    Message  string    `json:"message,omitempty"`
+    Location *FailedAt `json:"location,omitempty"`
+}
+``` 
+
+Second, we keep the `failedAt` field to prevent breaking user applications. However, it should be deprecated and removed in the future due to its redundancy. Therefore, we propose adding documents that state the deprecation status and adding the news to the release notes.
+To ease deprecation and future removal, we propose refactoring `UpdateBuildRunUsingTaskRunCondition` by extracting the logic of finding the failed pod and container in a BuildRun/TaskRun.
+The new file `pkg/reconciler/buildrun/resources/failures.go` includes the extraced logic and other failure related functions.
+
+A TaskRun can emit failures using two results:
+1. shp-error-reason
+2. shp-error-message
+
+Whenever a Tekton TaskRun fails, it keeps a JSON string of all results inside each terminated step in `TaskRun.Status.Steps[i].Terminated.Message` of type `PipelineResourceResult`. Whenever we can find the two error results in a step's message field, we surface the error as `failureDetails`.
+
+So far, we have presented a mechanism to lift errors results into a BuildRun. However, no component has produced error results yet.  One source of failure is misconfiguration in the source step for git applications. For example, a user can forget to provide authentication details for a private repository. Therefore, we propose to implement error results for `cmd/git`. On top of that, we propose implementing smarter error reporting by parsing and classifying git's messages for errors in `stdout`. A third-party application would be able to provide mitigations for failed builds based on error classes in git, for example, a UI workflow to reenter basic authentication details.
+
+The following errors classes are implemented for git:
+- `AuthInvalidUserOrPass`, expresses that basic authentication is not possible
+- `AuthInvalidKey`, expresses that ssh authentication is not possible
+- `BranchNotFound` expresses that a remote branch does not exist
+- `RepositoryNotFound`, expresses that the remote target for the git operation does not exist. It triggers when an error message is enough to determine that the remote target does not exist and is mostly derived from the Git server's messages e.g. GitLab or GitHub
+- `AuthPrompted`, is caused when a repo is not found, is private and authentication is insufficient
+- `Unknown`, is the class of choice if no other class fits
```

Comment:
When surfacing the reason code, I would prepend `Git` to each one:
```suggestion
- `GitAuthInvalidUserOrPass`, expresses that basic authentication is not possible
- `GitAuthInvalidKey`, expresses that ssh authentication is not possible
- `GitBranchNotFound` expresses that a remote branch does not exist
- `GitRepositoryNotFound`, expresses that the remote target for the git operation does not exist. It triggers when an error message is enough to determine that the remote target does not exist and is mostly derived from the Git server's messages e.g. GitLab or GitHub
- `GitAuthPrompted`, is caused when a repo is not found, is private and authentication is insufficient
- `GitUnknown`, is the class of choice if no other class fits
```

Reply:
I saw the way that you used it in the condition example above. Generally these enums are only for referencing the reason throughout the code. Since the package is already called `git`, I don't see the value in including git in the enum names. For presentation, the error classes implement the `Stringers` interface. There I have included `git` as prefix e.g. in https://github.com/dalbar/build/blob/add/surface_error_results/docs/buildrun.md?plain=1#L300. 

Would that be sufficient for you?